xbmc/platform/posix/PosixTimezone.cpp

```diff
@@ -157,85 +242,40 @@ void CPosixTimezone::OnSettingChanged(const std::shared_ptr<const CSetting>& set
 void CPosixTimezone::OnSettingsLoaded()
 {
   SetTimezone(CServiceBroker::GetSettingsComponent()->GetSettings()->GetString(CSettings::SETTING_LOCALE_TIMEZONE));
-  CDateTime::ResetTimezoneBias();
 }
 
-std::vector<std::string> CPosixTimezone::GetCounties()
+std::vector<std::string> CPosixTimezone::GetCountries()
 {
-   return m_counties;
+  return m_countries;
 }
 
 std::vector<std::string> CPosixTimezone::GetTimezonesByCountry(const std::string& country)
 {
-   return m_timezonesByCountryCode[m_countryByName[country]];
+  return m_timezonesByCountryCode[m_countryByName[country]];
 }
 
 std::string CPosixTimezone::GetCountryByTimezone(const std::string& timezone)
 {
 #if defined(TARGET_DARWIN)
-   return "?";
-#else
-   return m_countryByCode[m_countriesByTimezoneName[timezone]];
+  return "?";
 #endif
+
+  return m_countryByCode[m_countriesByTimezoneName[timezone]];
 }
 
 void CPosixTimezone::SetTimezone(const std::string& timezoneName)
 {
-#if !defined(TARGET_DARWIN)
-  bool use_timezone = true;
-#else
-  bool use_timezone = false;
+#if defined(TARGET_DARWIN)
+  return;
 #endif
 
-  if (use_timezone)
-  {
-    static char env_var[255];
-    sprintf(env_var, "TZ=:%s", timezoneName.c_str());
-    putenv(env_var);
-    tzset();
-  }
+  setenv("TZ", timezoneName.c_str(), 1);
+  tzset();
 }
 
 std::string CPosixTimezone::GetOSConfiguredTimezone()
 {
-   char timezoneName[255];
-
-   // try Slackware approach first
-   ssize_t rlrc = readlink("/etc/localtime-copied-from"
-                           , timezoneName, sizeof(timezoneName)-1);
-
-   // RHEL and maybe other distros make /etc/localtime a symlink
-   if (rlrc == -1)
-     rlrc = readlink("/etc/localtime", timezoneName, sizeof(timezoneName)-1);
-
-   if (rlrc != -1)
-   {
-     timezoneName[rlrc] = '\0';
-
-     char* p = strrchr(timezoneName,'/');
-     if (p)
-     { // we want the previous '/'
-       char* q = p;
-       *q = 0;
-       p = strrchr(timezoneName,'/');
-       *q = '/';
-       if (p)
-         p++;
-     }
-     return p;
-   }
-
-   // now try Debian approach
-   timezoneName[0] = 0;
-   FILE* fp = fopen("/etc/timezone", "r");
-   if (fp)
-   {
-      if (fgets(timezoneName, sizeof(timezoneName), fp))
-        timezoneName[strlen(timezoneName)-1] = '\0';
-      fclose(fp);
-   }
-
-   return timezoneName;
+  return date::current_zone()->name();
```

Comment:
This gets called before the call to `date::set_install(tzdataPath);` in Application.cpp. This causes an unhandled exception to be thrown. Trace:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Timezone database not found at "/home/mark/Downloads/tzdata"

Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
44            return INTERNAL_SYSCALL_ERROR_P (ret) ? INTERNAL_SYSCALL_ERRNO (ret) : 0;
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007ffff6e836b3 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007ffff6e33958 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff6e1d53d in __GI_abort () at abort.c:79
#4  0x00007ffff5699833 in __gnu_cxx::__verbose_terminate_handler () at /usr/src/debug/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#5  0x00007ffff56a5cfc in __cxxabiv1::__terminate (handler=<optimized out>) at /usr/src/debug/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:48
#6  0x00007ffff56a5d69 in std::terminate () at /usr/src/debug/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:58
#7  0x00007ffff56a5fcd in __cxxabiv1::__cxa_throw (obj=<optimized out>, tinfo=0x7ffff58291e0 <typeinfo for std::runtime_error>, dest=0x7ffff56bd470 <std::runtime_error::~runtime_error()>) at /usr/src/debug/gcc/libstdc++-v3/libsupc++/eh_throw.cc:98
#8  0x000055556880697e in date::init_tzdb() ()
#9  0x00005555687fd2bf in date::create_tzdb() ()
#10 0x00005555687fd37b in date::get_tzdb_list() ()
#11 0x0000555568807218 in date::get_tzdb() ()
#12 0x0000555568808995 in date::current_zone() ()
#13 0x000055556132d234 in CPosixTimezone::GetOSConfiguredTimezone[abi:cxx11]() (this=0x55556e245940 <g_timezone>) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/platform/posix/PosixTimezone.cpp:278
#14 0x0000555565bd8fb5 in CSettings::InitializeDefaults (this=0x60d000002810) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/settings/Settings.cpp:740
#15 0x0000555565bd563f in CSettings::InitializeDefinitions (this=0x60d000002810) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/settings/Settings.cpp:691
#16 0x0000555565bce029 in CSettings::Initialize (this=0x60d000002810) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/settings/Settings.cpp:473
#17 0x0000555565c21470 in CSettingsComponent::Initialize (this=0x608000023a30) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/settings/SettingsComponent.cpp:59
#18 0x00005555645103ac in CAppEnvironment::SetUp (appParams=warning: RTTI symbol not found for class 'std::_Sp_counted_ptr_inplace<CAppParams, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>'
warning: RTTI symbol not found for class 'std::_Sp_counted_ptr_inplace<CAppParams, std::allocator<void>, (__gnu_cxx::_Lock_policy)2>'
std::shared_ptr<CAppParams> (use count 4, weak count 0) = {...}) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/application/AppEnvironment.cpp:20
#19 0x000055556048041c in main (argc=1, argv=0x7fffffffe498) at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/platform/posix/main.cpp:70
```
So either ensure the initialization happens first or handle the exception here.

Reply:
I think this function should be left intact. It is called only by Settings and its main purpose is to get timezone before bundled tzdb is loaded. Thanks for pointing this!